### DISCUSSION TITLE
How to tell if dragon fruit is ripe using computer vision?

### DISCUSSION THREAD

**Discussion by NLadi** [ultralytics/ultralytics#22138] [Show and tell]
Traditional cues like skin color or firmness are often unreliable when it comes to judging dragon fruit ripeness. Enter computer vision.

In our latest blog, Abirami Vina explores how computer vision models like Ultralytics YOLO11 help farmers and consumers detect ripeness with speed, accuracy, and consistency, whether through drones, sorting systems, or even harvesting robots.

Learn more ➡️ https://www.ultralytics.com/blog/how-to-tell-if-dragon-fruit-is-ripe-using-computer-vision

[![white-fleshed-pitahaya-thai-dragon-fruit-in-stain-2025-03-09-10-41-45-utc (1)](https://github.com/user-attachments/assets/2c64f31f-3336-4a72-b1fd-c13d04492236)](https://www.ultralytics.com/blog/how-to-tell-if-dragon-fruit-is-ripe-using-computer-vision)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @NLadi, thank you for your interest in Ultralytics 🚀! This is an automated response—an Ultralytics engineer will also assist you here soon. In the meantime, we recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Thanks for sharing, @NLadi—great topic. For anyone wanting to prototype this, start with a simple 3‑class YOLO11 image classifier (underripe/ripe/overripe) for packhouse images, then move to detection/segmentation for in‑field occlusions; the dataset should follow the standard folder layout in the [classification task guide](https://docs.ultralytics.com/tasks/classify/) and you can find labeling tips in the [data collection and annotation guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/). 

```python
from ultralytics import YOLO

# 3-class ripeness classifier
model = YOLO('yolo11n-cls.pt')
model.train(data='data/dragonfruit-cls', epochs=50, imgsz=224, batch=64)
model.predict(source='data/dragonfruit-cls/val', save=True)
```

If you need orchard‑scale counting or overlapping fruit, switch to detection with a small YAML per the [detection task guide](https://docs.ultralytics.com/tasks/detect/).
